### PR TITLE
Task/issue 9 parsing perf improvements

### LINKED ISSUE #9: Performance enhancements around serialization
I ran some performance tests comparing Manatee.Json to Newtonsoft, and Manatee lost. Badly.

There are opportunities for making Manatee.Json (much) faster. This task is for finding and implementing those improvements.

As of publishing the parsing improvements:
- Newtonsoft beats Manatee in tests with 500+ iterations.
- Manatee prevails with less than 500 (singles are best).

It seems that Newtonsoft has some sort of caching mechanism that helps with rapid repeated serialization.

At 10K iterations, Manatee takes 10x the time of Newtonsoft. (This is down from ~50x.) The reverse is true for single iterations (Manatee takes 1/10 the time of Newtonsoft).

### PR DESCRIPTION
**Pull request by gregsdennis**
Resolves #9.

This PR consolidates all of the awesome effort put forth by @sixlettervariables in regard to performance enhancements.  This branch is continually tested and will be merged once we deem that the enhancements are as complete as possible.

### CODE REVIEW

File: Manatee.Json/Serialization/AbstractionMap.cs [hunk 1]
```diff
@@ -94,17 +94,22 @@ public void RemoveMap<TAbstract>(bool removeRelated = true)
 		/// <returns>The mapped type if a mapping exists; otherwise the abstraction type.</returns>
 		public Type GetMap(Type type)
 		{
-			if (!type.GetTypeInfo().IsAbstract && !type.GetTypeInfo().IsInterface) return type;
 			if (_registry.TryGetValue(type, out Type tConcrete)) return tConcrete;
-
+			if (!type.GetTypeInfo().IsAbstract && !type.GetTypeInfo().IsInterface)
```
Comment (gregsdennis) [inline Manatee.Json/Serialization/AbstractionMap.cs]: @sixlettervariables, this is breaking two tests: 

- `JsonSerializerTests.MapGenericAbstraction_BaseClass_Success()`
- `JsonSerializerTests.MapGenericAbstraction_WithOverride_Success()`

The purpose of these two tests is to verify that a non-abstract base class registration can be mapped to a derived class.  The generic type has to be checked before we can add it to the registry.

Comment (gregsdennis) [reply Manatee.Json/Serialization/AbstractionMap.cs]: I've moved it to after the generic check.

**Comment by gregsdennis** [comment]
Oh!  Forgot I commented out the circular reference tests... working on those.

**Comment by gregsdennis** [comment]
There was nothing wrong with the circular serialization.  The test object was recursing in `.GetHashCode()` due to the introduction of `SerializationPairCache._objectCache`.

**Comment by gregsdennis** [comment]
Anecdotally, these changes have improved the included test (see Console project) from 5m12s (on master) to 2m09s.

**Comment by sixlettervariables** [comment]
Good to know. I've not really thoroughly tested these changes, as I've mainly been evaluating it within our product. NUnit was failing with some bizarre errors today (regarding .NET Standard 1.3), but now that I am basically done with my updates I'll take a look at that situation.

**Comment by gregsdennis** [comment]
For the record, that 2 minutes to parse & deserialize 500,000 of these:

    {
      "_id": "5a2a6507e054a829a2e9ca74",
      "index": 0,
      "guid": "1469f18c-32c0-47cb-b108-328357f5501b",
      "isActive": false,
      "balance": "$1,575.19",
      "picture": "http://placehold.it/32x32",
      "age": 24,
      "eyeColor": "blue",
      "name": "Wiley Cunningham",
      "gender": "male",
      "company": "SEQUITUR",
      "email": "wileycunningham@sequitur.com",
      "phone": "+1 (943) 517-2400",
      "address": "968 Knapp Street, Shrewsbury, Texas, 6014",
      "about": "Voluptate excepteur voluptate eiusmod tempor culpa qui veniam laboris laboris officia in consectetur ea. Officia irure do magna eiusmod. Cillum in proident voluptate magna duis ad occaecat veniam in pariatur consectetur do. Cupidatat ullamco sint ex ipsum.\r\n",
      "registered": "2016-11-10T02:35:57 -13:00",
      "latitude": -24.154003,
      "longitude": -80.464485,
      "tags": [
        "voluptate",
        "et",
        "eu",
        "pariatur",
        "laboris",
        "exercitation",
        "cillum"
      ],
      "friends": [
        {
          "id": 0,
          "name": "Jayne Koch"
        },
        {
          "id": 1,
          "name": "Vera Humphrey"
        },
        {
          "id": 2,
          "name": "Jenifer Jarvis"
        }
      ],
      "greeting": "Hello, Wiley Cunningham! You have 2 unread messages.",
      "favoriteFruit": "strawberry"
    }

**Comment by sixlettervariables** [comment]
Aside: I've got NUnit working. Turned out to be unrelated to .NET Standard, just a cryptic error message due to an old extension.

File: Manatee.Json/Internal/ObjectCache.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+					}
+				}
+			}
+
+			return _builder();
+		}
+
+		private void _ReleaseSlow(T obj)
+		{
+			for (int ii = 0; ii < _items.Length; ii++)
```
Comment (gregsdennis) [inline Manatee.Json/Internal/ObjectCache.cs]: @sixlettervariables why do you use `ii`?

**Comment by sixlettervariables** [comment]
I've always doubled my single letter variables, but can switch them out to match the existing style (I honestly didn't notice).

**Comment by gregsdennis** [comment]
No worries, I've been changing them as I find them.  Really just curious.

**Comment by sixlettervariables** [comment]
As a note, I believe I'm finished as far as what I'd hoped to accomplish for performance improvements.

**Comment by gregsdennis** [comment]
Fantastic. Your help is greatly appreciated!

I'll go over all the changes again today and move toward merging into master.

**Comment by gregsdennis** [comment]
@sixlettervariables publishing v9.5.2.  Thanks for your help.  Not sure if you noticed, but I added you to the readme.

**Comment by sixlettervariables** [comment]
I appreciate that! Thank you for building a solid foundation upon which to improve.

**Pull request merged**